### PR TITLE
[scripts] Rig combine_ali_dirs.sh to combine alignment lattices

### PR DESCRIPTION
**Pull request by kkm000**
Relevant discussion:
https://groups.google.com/forum/#!topic/kaldi-help/2uxfByEAmfw

### CODE REVIEW

**Comment by danpovey** [comment]
Cool.  How much testing have you done?

**Comment by kkm000** [comment]
I merged ali and lats from two and three directories, and then extracted scps from 11-way split .gz arks and verified (by eyeballing) that the split was on a speaker boundary, and that the total numbers matched. I did not run failure cases though (with missing alignments). Can do that.

File: egs/wsj/s5/steps/combine_ali_dirs.sh [hunk 1]
```diff
@@ -1,105 +1,209 @@
...
+  # Assign all source gzipped archive names to an exported variable, one each
+  # per source directory, so that we can copy archives in a job per source.
+  src_id=0
+  for src in $@; do
+    let src_id++
+    nj_src=$(cat $src/num_jobs) || exit 1
+    # Create and export variable src_arcs_${src_id} for the job runner.
+    # Each numbered variable will contain the list of archives, e. g.:
+    # src_arcs_1="exp/tri3_ali/ali.1.gz exp/tri3_ali/ali.1.gz ..."
+    # ('printf' repeats its format as long as there are more arguments).
```
Comment (danpovey) [inline egs/wsj/s5/steps/combine_ali_dirs.sh]: I don't see the -v option for printf on my mac.

Comment (kkm000) [reply egs/wsj/s5/steps/combine_ali_dirs.sh]: Should be bash 3.2? I try to verify all stuff on docker with it, checks out here. Could you please try if `printf -v foo 'bar'` works?
```
$ docker run --rm -it bash:3.2
bash-3.2# printf -v foo 'bar'
bash-3.2# echo $foo
bar
```

**Comment by kkm000** [comment]
Tested with some alignment missing too (the original half-sets are both 16-job):
```
$ steps/combine_ali_dirs.sh --nj 11 data/train/vcogs/train_650 exp/vcogs-sr1/chain/ali_train_650 exp/vcogs-sr0.70/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh --nj 11 data/train/vcogs/train_650 exp/vcogs-sr1/chain/ali_train_650 exp/vcogs-sr0.70/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh: note: Temporary directory exp/vcogs-sr1/chain/ali_train_650/temp.epruSO will not be deleted in case of script failure, so you could examine it for troubleshooting.
steps/combine_ali_dirs.sh: Gathering alignments from each source directory.
steps/combine_ali_dirs.sh: Splitting combined alignments into 11 archives on speaker boundary.
steps/combine_ali_dirs.sh: Combined 342468 alignments for 342468 utterances.
steps/combine_ali_dirs.sh: There were 0 alignments not matching any utterance.
steps/combine_ali_dirs.sh: There were 0 utterances (0%) without alignments.
steps/combine_ali_dirs.sh: Gathering lattices from each source directory.
steps/combine_ali_dirs.sh: Splitting combined lattices into 11 archives on speaker boundary.
steps/combine_ali_dirs.sh: Combined 342468 lattices for 342468 utterances.
steps/combine_ali_dirs.sh: There were 0 lattices not matching any utterance.
steps/combine_ali_dirs.sh: There were 0 utterances (0%) without lattices.
steps/combine_ali_dirs.sh: Stored combined alignments and lattices in exp/vcogs-sr1/chain/ali_train_650
$ echo 15 >exp/vcogs-sr1.00/chain/ali_train_650/num_jobs
$ steps/combine_ali_dirs.sh --nj 11 --combine-lat false data/train/vcogs/train_650 exp/vcogs-sr1/chain/ali_train_650 exp/vcogs-sr0.70/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh --nj 11 --combine-lat false data/train/vcogs/train_650 exp/vcogs-sr1/chain/ali_train_650 exp/vcogs-sr0.70/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh: note: Temporary directory exp/vcogs-sr1/chain/ali_train_650/temp.2yVs5W will not be deleted in case of script failure, so you could examine it for troubleshooting.
steps/combine_ali_dirs.sh: Gathering alignments from each source directory.
steps/combine_ali_dirs.sh: Splitting combined alignments into 11 archives on speaker boundary.
steps/combine_ali_dirs.sh: Combined 342468 alignments for 331766 utterances.
steps/combine_ali_dirs.sh: There were 0 alignments not matching any utterance.
steps/combine_ali_dirs.sh: There were 10702 utterances (3%) without alignments.
steps/combine_ali_dirs.sh: Stored combined alignments in exp/vcogs-sr1/chain/ali_train_650
```

**Comment by danpovey** [comment]
Would you mind combining the diagnostics so they take up fewer lines?
My bash version:

bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin18)
Copyright (C) 2007 Free Software Foundation, Inc.

printf -v foo 'bar' prints nothing for me on mac

Dan

File: egs/wsj/s5/steps/combine_ali_dirs.sh [hunk 2]
```diff
@@ -1,105 +1,209 @@
...
+do_combine() {
+  local ark=$1 entities=$2 copy_program=$3
+  shift 3
+
+  echo "$0: Gathering $entities from each source directory."
+  # Assign all source gzipped archive names to an exported variable, one each
+  # per source directory, so that we can copy archives in a job per source.
+  src_id=0
+  for src in $@; do
+    let src_id++
```
Comment (danpovey) [inline egs/wsj/s5/steps/combine_ali_dirs.sh]: can you just do `src_id=$[src_id+1]`, which  everyone knows?  Sometimes you are a little *too* knowledgeable.

**Comment by kkm000** [comment]
> Would you mind combining the diagnostics so they take up fewer lines?

You mean the last 3 lines? Sure.

> printf -v foo 'bar' prints nothing for me on mac

That's a good sign :) And 'echo $foo' prints 'bar' after it, just to be completely sure nothing weird is going on?

> GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin18)

My docker image for bash 3.2 has the same version. Weird, but its help does mention -v

```
bash-3.2# help printf
printf: printf [-v var] format [arguments]
    printf formats and prints ARGUMENTS under control of the FORMAT. FORMAT
    is a character string which contains three types of objects: plain
    characters, which are simply copied to standard output, character escape
    sequences which are converted and copied to the standard output, and
    format specifications, each of which causes printing of the next successive
    argument.  In addition to the standard printf(1) formats, %b means to
    expand backslash escape sequences in the corresponding argument, and %q
    means to quote the argument in a way that can be reused as shell input.
    If the -v option is supplied, the output is placed into the value of the
    shell variable VAR rather than being sent to the standard output.
bash-3.2# bash --version
GNU bash, version 3.2.57(1)-release (x86_64-pc-linux-musl)
Copyright (C) 2007 Free Software Foundation, Inc.
```

File: egs/wsj/s5/steps/combine_ali_dirs.sh (hunk 2 above)
Comment (kkm000) [reply egs/wsj/s5/steps/combine_ali_dirs.sh]: Is `src_id=$((src_id+1))` ok? The `$[...]` syntax for arithmetic is deprecated as of bash 4.

**Comment by danpovey** [comment]
Yes, (()) is OK.
RE the -v thing, yes it does print bar.  But remember we don't have an
explicit version dependency with bash (AFAIK), so we shouldn't use
version-dependent features, within reason of course.

**Comment by kkm000** [comment]
> we don't have an explicit version dependency with bash

I do not think bash earlier than 3.2 exists in the wild any more, and 3.2 is only on the Mac. And configure checks for it.

```
[[ $BASH_VERSION < '3.2' ]] && {
  echo >&2 "bash version ${BASH_VERSION} is too old, cannot continue." \
           "You won't be able to run Kaldi recipes with it anyway." \
           "Please upgrade. bash version 3.2 or higher is required."
  exit 1;
}
```

**Comment by danpovey** [comment]
But please work with this for a while, and remind me in a week or so, I don't think it's super urgent.  I'd rather wait and see if any problem comes up than merge right away.

**Comment by kkm000** [comment]
This script is actually used only once in the whole egs/ dir:
```
$ find egs/ -type f | xargs grep 'combine_ali_dirs.sh'
egs/aspire/s5/local/multi_condition/run_nnet2_common.sh:  steps/combine_ali_dirs.sh data/train_rvb exp/tri5a_rvb_ali $ali_dirs || exit 1;
```
Not very low-level.

**Comment by kkm000** [comment]
Arithmetics in `$(( ))`, diagnostic made more compact:
```
$ steps/combine_ali_dirs.sh --nj 16 data/train/vcogs8/train_650 exp/vcogs-temp/chain/ali_train_650 exp/vcogs-sr0.80/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh --nj 16 data/train/vcogs8/train_650 exp/vcogs-temp/chain/ali_train_650 exp/vcogs-sr0.80/chain/ali_train_650 exp/vcogs-sr1.00/chain/ali_train_650
steps/combine_ali_dirs.sh: note: Temporary directory exp/vcogs-temp/chain/ali_train_650/temp.AFs6Gb will not be deleted in case of script failure, so you could examine it for troubleshooting.
steps/combine_ali_dirs.sh: Gathering alignments from each source directory.
steps/combine_ali_dirs.sh: Splitting combined alignments into 16 archives on speaker boundary.
steps/combine_ali_dirs.sh: Combined 342468 alignments for 342468 utterances. There were 0 utterances (0%) without alignments, and 0 alignments not matching any utterance.
steps/combine_ali_dirs.sh: Gathering lattices from each source directory.
steps/combine_ali_dirs.sh: Splitting combined lattices into 16 archives on speaker boundary.
steps/combine_ali_dirs.sh: Combined 342468 lattices for 342468 utterances. There were 0 utterances (0%) without lattices, and 0 lattices not matching any utterance.
steps/combine_ali_dirs.sh: Stored combined alignments and lattices in exp/vcogs-temp/chain/ali_train_650
```

@danpovey, I'll be using this script a few times during the week. We can let it stew for a while to make sure, if you prefer. Only all its uses are going to be boringly similar here, not the best for testing.

**Comment by danpovey** [comment]
OK thanks, will merge now.

**Pull request merged**